File: bean-validation/src/test/java/io/github/krandom/validation/FutureAnnotationHandlerTest.java
package io.github.krandom.validation;

import static org.assertj.core.api.Assertions.assertThat;

import io.github.krandom.KRandom;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import jakarta.validation.constraints.Future;
import java.time.*;
import java.time.chrono.HijrahDate;
import java.time.chrono.JapaneseDate;
import java.time.chrono.MinguoDate;
import java.time.chrono.ThaiBuddhistDate;
import java.util.Calendar;
import java.util.Date;
import java.util.Set;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class FutureAnnotationHandlerTest {

  private KRandom kRandom;

  @BeforeEach
  void setUp() {
    kRandom = new KRandom();
  }

  @Test
  void generatedBeanShouldBeValidAccordingToValidationConstraints() {
    TestBean testBean = kRandom.nextObject(TestBean.class);

    assertThat(testBean.getTestDate()).isAfter(Date.from(Instant.now()));
    assertThat(testBean.getTestCalendar()).isGreaterThan(Calendar.getInstance());
    assertThat(testBean.getTestInstant()).isAfter(Instant.now());
    assertThat(testBean.getTestLocalDate()).isAfter(LocalDate.now());
    assertThat(testBean.getTestLocalDateTime()).isAfter(LocalDateTime.now());
    assertThat(testBean.getTestLocalTime()).isAfter(LocalTime.now(ZoneId.systemDefault()));
    assertThat(testBean.getTestMonthDay()).isGreaterThan(MonthDay.now());
    assertThat(testBean.getTestOffsetDateTime()).isAfter(OffsetDateTime.now());
    assertThat(testBean.getTestOffsetTime()).isAfter(OffsetTime.now());
    assertThat(testBean.getTestYear()).isGreaterThan(Year.now());
    assertThat(testBean.getTestYearMonth()).isAfter(YearMonth.now());
    assertThat(testBean.getTestZonedDateTime()).isAfter(ZonedDateTime.now());
    assertThat(testBean.getTestHijrahDate()).isGreaterThan(HijrahDate.now());
    assertThat(testBean.getTestJapaneseDate()).isGreaterThan(JapaneseDate.now());
    assertThat(testBean.getTestMinguoDate()).isGreaterThan(MinguoDate.now());
    assertThat(testBean.getTestThaiBuddhistDate()).isGreaterThan(ThaiBuddhistDate.now());
  }

  @Test
  void generatedBeanShouldBeValidUsingBeanValidationApi() {
    TestBean testBean = kRandom.nextObject(TestBean.class);

    Validator validator;
    try (ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory()) {
      validator = validatorFactory.getValidator();
    }
    Set<ConstraintViolation<TestBean>> violations = validator.validate(testBean);

    assertThat(violations).isEmpty();
  }

  static class TestBean {
    @Future private Date testDate;

    public Date getTestDate() {
      return testDate;
    }

    @Future private Calendar testCalendar;

    public Calendar getTestCalendar() {
      return testCalendar;
    }

    @Future private Instant testInstant;

    public Instant getTestInstant() {
      return testInstant;
    }

    @Future private LocalDate testLocalDate;

    public LocalDate getTestLocalDate() {
      return testLocalDate;
    }

    @Future private LocalDateTime testLocalDateTime;

    public LocalDateTime getTestLocalDateTime() {
      return testLocalDateTime;
    }

    @Future private LocalTime testLocalTime;

    public LocalTime getTestLocalTime() {
      return testLocalTime;
    }

    @Future private MonthDay testMonthDay;

    public MonthDay getTestMonthDay() {
      return testMonthDay;
    }

    @Future private OffsetDateTime testOffsetDateTime;

    public OffsetDateTime getTestOffsetDateTime() {
      return testOffsetDateTime;
    }

    @Future private OffsetTime testOffsetTime;

    public OffsetTime getTestOffsetTime() {
      return testOffsetTime;
    }

    @Future private Year testYear;

    public Year getTestYear() {
      return testYear;
    }

    @Future private YearMonth testYearMonth;

    public YearMonth getTestYearMonth() {
      return testYearMonth;
    }

    @Future private ZonedDateTime testZonedDateTime;

    public ZonedDateTime getTestZonedDateTime() {
      return testZonedDateTime;
    }

    @Future private HijrahDate testHijrahDate;

    public HijrahDate getTestHijrahDate() {
      return testHijrahDate;
    }

    @Future private JapaneseDate testJapaneseDate;

    public JapaneseDate getTestJapaneseDate() {
      return testJapaneseDate;
    }

    @Future private MinguoDate testMinguoDate;

    public MinguoDate getTestMinguoDate() {
      return testMinguoDate;
    }

    @Future private ThaiBuddhistDate testThaiBuddhistDate;

    public ThaiBuddhistDate getTestThaiBuddhistDate() {
      return testThaiBuddhistDate;
    }
  }
}

File: bean-validation/src/test/kotlin/io/github/krandom/validation/FutureAnnotationHandlerTest.kt
package io.github.krandom.validation

import io.github.krandom.KRandom
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.comparables.shouldBeGreaterThan
import io.kotest.matchers.nulls.shouldNotBeNull
import jakarta.validation.Validation
import jakarta.validation.Validator
import jakarta.validation.constraints.Future
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.MonthDay
import java.time.OffsetDateTime
import java.time.OffsetTime
import java.time.Year
import java.time.YearMonth
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.chrono.HijrahDate
import java.time.chrono.JapaneseDate
import java.time.chrono.MinguoDate
import java.time.chrono.ThaiBuddhistDate
import java.util.Calendar
import java.util.Date
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class FutureAnnotationHandlerTest {
  private lateinit var kRandom: KRandom

  @BeforeEach
  fun setUp() {
    kRandom = KRandom()
  }

  @Test
  fun generatedBeanShouldBeValidAccordingToValidationConstraints() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    testBean.shouldNotBeNull()
    testBean.testDate shouldBeGreaterThan Date.from(Instant.now())
    testBean.testCalendar shouldBeGreaterThan Calendar.getInstance()
    testBean.testInstant shouldBeGreaterThan Instant.now()
    testBean.testLocalDate shouldBeGreaterThan LocalDate.now()
    testBean.testLocalDateTime shouldBeGreaterThan LocalDateTime.now()
    testBean.testLocalTime shouldBeGreaterThan LocalTime.now(ZoneId.systemDefault())
    testBean.testMonthDay shouldBeGreaterThan MonthDay.now()
    testBean.testOffsetDateTime shouldBeGreaterThan OffsetDateTime.now()
    testBean.testOffsetTime shouldBeGreaterThan OffsetTime.now()
    testBean.testYear shouldBeGreaterThan Year.now()
    testBean.testYearMonth shouldBeGreaterThan YearMonth.now()
    testBean.testZonedDateTime shouldBeGreaterThan ZonedDateTime.now()
    testBean.testHijrahDate shouldBeGreaterThan HijrahDate.now()
    testBean.testJapaneseDate shouldBeGreaterThan JapaneseDate.now()
    testBean.testMinguoDate shouldBeGreaterThan MinguoDate.now()
    testBean.testThaiBuddhistDate shouldBeGreaterThan ThaiBuddhistDate.now()
  }

  @Test
  fun generatedBeanShouldBeValidUsingBeanValidationApi() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    val validator: Validator
    Validation.buildDefaultValidatorFactory().use { validatorFactory ->
      validator = validatorFactory.validator
    }
    val violations = validator.validate(testBean)

    violations.shouldBeEmpty()
  }

  internal data class TestBean(
    @field:Future val testDate: Date,
    @field:Future val testCalendar: Calendar,
    @field:Future val testInstant: Instant,
    @field:Future val testLocalDate: LocalDate,
    @field:Future val testLocalDateTime: LocalDateTime,
    @field:Future val testLocalTime: LocalTime,
    @field:Future val testMonthDay: MonthDay,
    @field:Future val testOffsetDateTime: OffsetDateTime,
    @field:Future val testOffsetTime: OffsetTime,
    @field:Future val testYear: Year,
    @field:Future val testYearMonth: YearMonth,
    @field:Future val testZonedDateTime: ZonedDateTime,
    @field:Future val testHijrahDate: HijrahDate,
    @field:Future val testJapaneseDate: JapaneseDate,
    @field:Future val testMinguoDate: MinguoDate,
    @field:Future val testThaiBuddhistDate: ThaiBuddhistDate,
  )
}
